### ISSUE TITLE
How do I install this?

### ISSUE THREAD

**Issue by VetheonGames** [docker-mailserver/docker-mailserver#2621]
# Subject
Documentation unclear

## Description

I have *NO IDEA* where to start setting this up. Do I clone the repository? Do I pull the Docker image? Do I do something otherwise? I don't know. 

I have read through ***EVERY OUNCE*** of provided documentation, watched several youtube videos, and read articles that weren't written by the team/person behind this project yet I still have *no idea how to start*.

So, I just wanna know, how do I start using this? I follow the documentation step by step but it's a mess! It makes no sense! THe steps are out of order, you need the setup.sh but you don't need it at the same time?

Is this project abandoned and not functional anymore and I'm banging my head off my desk for no reason?

Or is the documentation convoluted and hard to understand on purpose?

It really should NOT be that hard to just ***list the commands we need in the order we need to run them like every other developer on the planet***

Please. Someone. Provide me a link to documentation that  was written by someone with a handle on how to write tutorials? Because everything I've found docs wise regarding this container and image is non-sensical.

**Comment by VetheonGames** [comment]
Nevermind. I've decided to just completely forego using this or Docker at all. I've learned just how terrible docker is when compared to the software it shamelessly rips off. 

Later docker nerds!

**Comment by polarathene** [comment]
## TL;DR

I'm not sure why you've had trouble getting started, we seem to cover this well?

I have provided a simple 3 step guide below (add a config, run two commands, done).

I care about our docs being better, so if you'd be so kind as to help me figure out a better way to communicate what you'd need to know, please provide more feedback :)

## Full response

> So, I just wanna know, how do I start using this? I follow the documentation step by step but it's a mess! It makes no sense!
> The steps are out of order, you need the `setup.sh` but you don't need it at the same time?

Hi, this project was in a much worse state a couple years back. I and other volunteers have contributed significant time to address this and make it better for users.

We'd appreciate your feedback to be more detailed so we could work through that frustration you went through and learn early on where you started from so that we can improve that experience for other users like yourself in future. It's a big project and volunteers only have so much time to spare, but our docs are important for us to get right too.

---

Regarding `setup.sh` it's a CLI helper with some commands for administration that makes it easier for the user to perform certain actions. You are most likely to use it to setup the first e-mail account to get the project running.

At a later point the bulk of the `setup.sh` functionality was moved into the container as a `setup` command that can be called with `docker exec -it <container-name> setup <command>` instead, so that you don't need to rely on an external `setup.sh` script. `setup.sh` is still preferred by some users, and while mostly a wrapper now, it does provide some extra functionality that's mostly useful if you don't have a container already running, or want to type just a little less.

---

> Is this project abandoned and not functional anymore and I'm banging my head off my desk for no reason?

Look at the project activity? You'll see I'm regularly contributing improvements / fixes and responding to both user issues and discussions and resolving those as best I can, as are other maintainers. We have reasonably frequent releases as well.

Why would you consider that abandoned because you had difficulty, while many others are not? I would say that while we'd love to improve what we can to make your experience better, that you cannot place all the blame on us? It seems that you're new to some technologies like Docker, and perhaps mail servers (and their specific DNS records) are new to you as well?

We cannot teach you everything you should know, our focus is to make setting up a mail-server easier and is where our effort is spent. Many users have different needs and environments, but we provide the project like other Docker projects, with the assumption that you know how to use Docker. Our project is a bit more complicated being a mail-server, with all the configuration you must handle on your end, along with us providing a single container bundling multiple services (but this is one of the reasons users choose our project in particular).

---

> Or is the documentation convoluted and hard to understand on purpose?

No.. I worked very hard actually to clean our docs up from an original Github Wiki that was full of broken links and outdated information. We now have docs that are in sync with the git project commits, that automatically deploy, and has a nice web interface with additional features that Github Wiki lacked.

I assure you the docs are in a much nicer state than they once were, and that I've sunk tens of hours into making them better myself. You as a user may take for granted what you see, but it's a lot more work than it looks. I was confused and frustrated like you and I had much worse to work with.

Instead of complaining and running away, I got informed myself and contributed back to the community for the benefit of others. You don't have to do what I have, but you could assist volunteers like myself who are eager to help the community, by letting us know where you get stuck and we can improve our docs from that point as we go.

It's otherwise entitled to believe our docs should be perfect, when you're not willing to engage with us to help improve them. They can only get better when we have users share the pain points. I can only assume you've not volunteered / contributed to anything yet, and have high expectations for an open-source project maintained by unpaid volunteers and provided to you for free? You're welcome to do it all yourself, or pay someone to make things easy for you?

---

> It really should NOT be that hard to just list the commands we need in the order we need to run them like every other developer on the planet

Lets look at the [README](https://github.com/docker-mailserver/docker-mailserver#usage) shall we?

1. Get `docker-compose.yml`, `mailserver.env`, `setup.sh` files.
   
   ```bash
   DMS_GITHUB_URL='https://raw.githubusercontent.com/docker-mailserver/docker-mailserver/master'
   wget "${DMS_GITHUB_URL}/docker-compose.yml"
   wget "${DMS_GITHUB_URL}/mailserver.env"
   wget "${DMS_GITHUB_URL}/setup.sh"
   chmod a+x ./setup.sh
   ```
   
   Don't forget to edit `docker-compose.yml` and `mailserver.env` to your needs. If you have absolutely no idea what that is and want to start with the bare minimum, don't even bother with `mailserver.env`, just use this `docker-compose.yml` instead:
   
   ```yml
   services:
     mailserver:
       image: docker.io/mailserver/docker-mailserver:latest
       container_name: mailserver
       # This should be set to your DNS A record that you have an MX record pointing to,
       # Usually it's hosted on the same domain you want to send/receive as, eg `user@example.com`
       # The 'mail' part is to point to your MTA (this container), and does not affect your mail address,
       # Do not mistaken it as `user@mail.example.com`
       hostname: mail.example.com
       # More information about the mail-server ports:
       # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
       # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
       ports:
         # Port 25, mail sent to your server from others comes in through this port.
         - "25:25"
         # All four ports below may only be relevant to you with a mail client like ThunderBird
         # These two ports are only relevant if you want to send mail from your server,
         # and that mail client submitting the mail to send needs to be authorized via login.
         - "465:465"
         - "587:587"
         # IMAP4 - Only relevant if you want to read mail sent to your sever,
         # You only need the port your mail client uses.
         - "143:143"
         - "993:993"
       # Persist data across container restarts by mounting these local folders:
       volumes:
         - ./docker-data/dms/mail-data/:/var/mail/
         - ./docker-data/dms/mail-state/:/var/mail-state/
         - ./docker-data/dms/mail-logs/:/var/log/mail/
         - ./docker-data/dms/config/:/tmp/docker-mailserver/

2. Blah blah, skip passed helpful advice to [this part](https://github.com/docker-mailserver/docker-mailserver#starting-for-the-first-time):

   ```bash
   ./setup.sh email add user@example.com this-is-a-weak-password
   docker-compose up -d
   ```

Congratulations, you have a mail-server up and running! You'll need to actually have DNS setup properly of course, and perhaps you have some other issues or expectations... You weren't very clear with your struggle beyond getting the project up and running..

Once more, just for the sake of clarity:

- Create a `docker-compose.yml` configured like above.
- Start it `docker-compose up -d`
- Didn't setup an account with `setup.sh` earlier? You have 2 minutes during startup to add one:
   `docker exec -it mailserver setup email add user@example.com this-is-a-weak-password`
- Done!

I assume that's not a sufficient list for you, so why not chime in with what it's missing?

---

> I have NO IDEA where to start setting this up. Do I clone the repository? Do I pull the Docker image? Do I do something otherwise? I don't know.
> 
> I have read through EVERY OUNCE of provided documentation, watched several youtube videos, and read articles that weren't written by the team/person behind this project yet I still have no idea how to start.

Really...? You didn't come across anything like shown above? I made some minor tweaks from the referenced README, but you'd have become aware of such changes if you really did read our docs.

---

> Please. Someone. Provide me a link to documentation that was written by someone with a handle on how to write tutorials? Because everything I've found docs wise regarding this container and image is non-sensical.

Most of the docs are community contributions. I've revised a fair bit of it, but I can totally understand that we could do better here. If you're willing to engage, we can make that happen together.

---

> Nevermind. I've decided to just completely forego using this or Docker at all. I've learned just how terrible docker is when compared to the software it shamelessly rips off.

What software do you think is being shamelessly ripped off here? We aren't pretending that we don't use third-party software. This project integrates all the software you'd need to do yourself (some you may not need, we cater to a broad audience).

You're welcome to setup Postfix yourself, plenty of guides out there. Depending on your needs that can be easier too. The Postfix docs can get confusing at times if you have to actually dig into them. If your needs are basic, this could definitely be better/easier for you.

Where it can get more problematic is when you need to integrate with other software, Dovecot, Fail2Ban, Amavis+ClamAV+SpamAssassin, making config changes (including when your services get updated certs), what TLS policies you should use, account management, and so forth..

Those tutorials start to get long-winded, or worse you're juggling several separate ones and trying to blend them together (worse if they're severely out of date). All you're managing to do at that point is basically what this project handles (except we unify as a community to maintain it), but you are then responsible for all maintenance and keeping on top of things.

---

If you don't understand the benefits of Docker, that's fine. I wish you good luck.

I'm a little disheartened that you only seem interested in being disgruntled and disrespecting all the effort the community puts into making this project better. If you change your mind and would like to assist with improving our docs, that'd be awesome! :grinning:

**Comment by VetheonGames** [comment]
For the record,

I actually apologize for how rude I was.

However, I have discovered how awful Docker is and have decided to simply use LXC for any containers I might ever need.

I despise and hate docker with a burning firey passion, yet my sysadmins continue to encourage me to learn it, despite it being syntaxed by a ten-year-old.

Before I read your rather in-depth reply, I actually *did* just setup a postfix, dovecot and roundcube setup by myself. And frankly, it was 1000x easier than installing this image was.

Like... for real... it took me 3 hours to have PostFix, Dovecot, and Roundcube running and properly working with OpenDKIM etc.

I don't know, maybe I'm just different but, IMO, docker is an overrated LXC clone with better marketing. Nothing more.

**Comment by polarathene** [comment]
Apologies for the verbosity. I wanted to reply to everything, but don't have the time to make it terser :sweat_smile: 

I sectioned the response off with headings for each topic. I understand if you don't have the time to read through or respond to it all :+1: 

---

Thanks for clearing up the expectation for docs. I would like to tackle that if I can find the time, but TLDR is I have to prioritize more pressing tasks while I can and rely on volunteers for our docs.

Could you let us know where you started with our docs?
- Did you see our [project `README`](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md)? If so did you have trouble follow instructions there to get a container running?
- Was there something wrong with [our main docs page](https://docker-mailserver.github.io/docker-mailserver/edge/)? Was something missing that should be there?

I can make some minor improvements to the docs. We could at least try ensure users don't have trouble starting the basic container setup. Larger restructuring and revisions will have to wait unfortunately.

---

> I actually apologize for how rude I was.
> And again, I do apologize for attacking you, you didn't deserve that. 

Thank you, I appreciate that :)

### DIY

> I actually _did_ just setup a postfix, dovecot and roundcube setup by myself.

Well done! But do keep biases in mind here.

You're experienced and more comfortable with setting all that up by yourself. Our users often come from a background where they are looking for a solution using Docker, and usually they're not comfortable figuring it all out by themselves.

Some could follow guides out there, but those can become outdated and may skip over some technical details that we address (_there's also a lot of bad advice out there, even from leading SMTP relay services docs, and they're paid+proprietary_ :grimacing: ). If you're comfortable maintaining your mail-server with a more hands-on approach, that's great too.

I won't say our project here is perfect by any means, abstractions do get in the way at times for sure. I do think we do a pretty good job though at providing a quick and easy to deploy mail-server, granted the docs for new users could be better as you've pointed out :)

> And frankly, it was 1000x easier than installing this image was.
> Like... for real... it took me 3 hours to have PostFix, Dovecot, and Roundcube running and properly working with OpenDKIM etc.

That's terrific, and I wish we could do better if that's the case.

The _1 config + 2 commands_ example I gave above would sort you out for Postfix + Dovecot (no need to spend 3 hours), and another command for OpenDKIM to add your domain(s). We've got Amavis + ClamAV + SpamAssassin covered too (the latter two need to be enabled via ENV), plus more depending on your needs.

RoundCube isn't something we directly support, but some users have setup a separate container alongside ours for that.

---

### Step-by-step guide

> But a word of advice, perhaps get someone to write a Step-By-Step for you that doesn't go into why everything is the way it is.

Oh we'd love that, but we're all volunteers. I've invested a tonne of time into the project but not made a cent from it. Ironically I don't even use it myself yet.

Someone has to come along and be willing to do this. I've made incremental steps to making that much easier, but I have higher priorities where the time I have available is better spent on the project currently. That's ensuring it's maintainability going forward, similar to what other current maintainers have been doing.

The project at an earlier stage was a disorganized mess evolving from a personal project with community contributions of varying quality tacked on over time. Docs will get another pass eventually, but as I don't know how long-term I can remain a maintainer on the project, it's important to me that it is easier to bring on new maintainers and contributors with stronger foundations to build upon. I don't say this lightly either, there have been some concerns that are very technical to tackle, requiring a fair chunk of time to implement properly, that is where I'm needed.

I doubt we can afford to pay anyone to work on the docs, and most users are not interested in contributing to the extent required. We get the odd improvement to our docs here and there, or someone writes a blog post externally that we add to our docs.

I do understand what you're asking for, providing a more clearer pathway to follow - beyond what our README details. It's on my TODO.

---

### LXC vs Docker

> I don't know, maybe I'm just different but, IMO, docker is an overrated LXC clone with better marketing. Nothing more.

Maybe? I cannot comment as I have only used Docker myself.

If I wanted to try LXC with a setup like we provide here, could you provide an easy to deploy equivalent, or would I need more knowledge vs the _1 config + 2 commands_ example I gave you above that works?

I think roughly it comes down to what you're comfortable with. Perhaps Docker itself was a bit of a added friction for you unrelated to anything about our project?

I don't know what issues I'd run into with LXC for the same reason I don't use Podman (very similar to Docker I hear, but it does have some subtle gotchas/differences). I assume I'd likewise run into some friction if I had the option to chose to give LXC a go, where I'd be faster at achieving the outcome I want with Docker, especially when it comes to third-party and collaboration with other devs.

---

### LinuxBabe Postfix setup article

> For an example of what I refer to by a proper tutorial, take a look at the guide I followed to setup PostFix:
> https://www.linuxbabe.com/redhat/run-your-own-email-server-centos-postfix-smtp-server

I know you'll probably not really read this, but I went over the linked article (1st part) and compared it to what this project does. You'd have gotten the bulk of that very quickly by just running the _1 config + 2 commands_ mentioned.

That is the point of what we're providing with this project. AFAIK, we provide sane defaults that likely match up with what the rest of that article series covers for you. Our docs touch on each specific area if it interests the user to deviate from the defaults we provide.

If you really did read all that you said you had, I'm not sure what wasn't covered for you? I do acknowledge that we could provide a better structured overview though, especially for DNS records and testing everything is working. We could really do with some more volunteers that can spare the time to work on that. I may be able to tackle it later this month.

#### Article comparison to DMS

> _Things To Do Before Installing Postfix_
> - _Set A Correct Hostname for CentOS 8/RHEL 8 Server_
> - _Create DNS Records for Your Mail Server_

You set the hostname part in the `docker-compose.yaml`, like in the example shown earlier to you.

DNS setup is definitely an area we could document better. We touch on individual records in pages specific to them, but could definitely add a page or section dedicated to that configuration step :+1: 
- I recently have been working on an [offline setup example](https://github.com/docker-mailserver/docker-mailserver/issues/2601#issuecomment-1135520225) to assist with troubleshooting.
- There's also [this guide in our docs](https://docker-mailserver.github.io/docker-mailserver/edge/examples/tutorials/basic-installation/) which does mention DNS `A`,`MX`,`TXT` (SPF and DKIM) records, with an example zonefile. I get that zonefiles may not be as friendly, and we could communicate the record config better :+1: 
- And our docs link to [this blogpost](https://www.ifthenel.se/self-hosted-mail-server/) which has a section on configuring DNS. DNS isn't specific to our project, being rather generic advice with mail servers, but I know this was a frustrating point for me when I got into it too.

> _Installing Postfix_

We take care of installing Postfix, that's shipped with the docker image (_it's just the same package you'd install from the base OS_).

> _Configuring Postfix_
> 1. _Listening on the public IP address_
> 2. _Setting the Postfix `hostname`, Setting `$mydomain` Parameter_
> 3. _Setting `$myorigin` Parameter, Setting $mydestination Parameter_
> 4. _Restarting Postfix_
> 5. _Open TCP Port 25 (inbound) in Firewall_
> 6. _Checking If Port 25 (outbound) is blocked_
> 7. _Testing mail with `sendmail` MUA - Sending Test Email_
> 8. _Inspecting logs - Still Can’t Send Email?_
> 9. _Using the `mail` MUA - Using the mail program to Send and Read Email On the Command Line_
> 10. _How To Increase Attachment Size Limit_
> 11. _Creating Email Alias_
> 12. _Using IPv4 Only_
> 13. _Postfix Automatic Restart_

1. `inet_interfaces = all` is [already configured](https://github.com/docker-mailserver/docker-mailserver/blob/0a722276a875ffd4ef7af1d16cbf516356222194/target/postfix/main.cf#L17).
2. `myhostname` and `mydomain` are [also configured just like your reference guide shows](https://github.com/docker-mailserver/docker-mailserver/blob/0a722276a875ffd4ef7af1d16cbf516356222194/target/scripts/startup/setup-stack.sh#L117-L118).
3. `myorigin` isn't set, but with our project that's a non-issue as this scenario doesn't apply (except for some internal services like Amavis, but these aren't sent external of your MTA). We have not configured `mydestination` to include `mydomain`, as we use `virtual_transport` with `virtual_mailbox_domains` for all user accounts to route mail to Dovecot over LMTP, not Postfix local delivery agent. [Official Postfix docs for `mydestination`](http://www.postfix.org/postconf.5.html#mydestination) advise not to mix such domains, so we're doing the correct thing there.
4. No restart is required, but when it is we handle that at the appropriate time (eg config overrides updated by user, we monitor such files for changes via a background service).
5. Firewall Port 25 inbound. This can depend on the OS, in some deployments you can just configure the ports as shown in `docker-compose.yml` and they'll be open to public routing to the containers port (Docker manages that). The [docs example guide for basic installation](https://docker-mailserver.github.io/docker-mailserver/edge/examples/tutorials/basic-installation/) does detail this with `ufw` instead of `firewalld` though.
6. Port 25 outbound is an issue we mention in the docs briefly; we could detail similar for how to check if it's blocked. Technically this is a prerequisite that should be checked before you start setting everything up on a host.
7. Sending a test email we could probably detail too if you think that's helpful with `sendmail` instead of having the user configure their preferred MUA (be that ThunderBird, or some separate service they want relay mail through the self-hosted MTA).
8. The mail logs you can get from the same location, but generally if you are testing your setup, your good with `docker-compose logs -f` (or just use `docker-compose up` without detached mode `-d`) and you'll get logs updating as they appear from services like Postfix. The logs can also be volume mounted, as shown in our `docker-compose.yml` example.
9. We could add a section for using the MUA `mail` like the article describes, but I don't think that our demographic would care for that vs a GUI like ThunderBird (or a webmail frontend).
10. `message_size_limit` is just another Postfix config setting. I don't know why, but [we have an ENV for that atm](https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/#postfix_message_size_limit).
    - This setting and others should really just be [handled with our config override support](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/override-defaults/postfix/), you include a `postfix-main.cf` file in your config volume mount, and that will append to `main.cf` (we filter out earlier entries by running it through `postconf`).
    - This is preferable over explicitly using `postconf` yourself to update `main.cf` as your config overrides can be version controlled for easy maintenance.
    - In a DIY, you need to perform this setup step each time, or some other way to apply it (considering package updates may overwrite `/etc/postfix/main.cf` or similar on some OS IIRC).
11. [We setup this alias for you](https://github.com/docker-mailserver/docker-mailserver/blob/0a722276a875ffd4ef7af1d16cbf516356222194/target/scripts/helpers/aliases.sh#L62), if you do not provide your own preferred account to alias to, [we set a default of `postmaster@<domain>` for you](https://github.com/docker-mailserver/docker-mailserver/blob/0a722276a875ffd4ef7af1d16cbf516356222194/target/scripts/start-mailserver.sh#L54) (eg: `postmaster@example.com` for the MX hostname `mail.example.com`).
12. Like Postfix, we don't know what defaults you want, you can configure this with the earlier mentioned `postfix-main.cf` file, or the [ENV `POSTFIX_INET_PROTOCOLS`](https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/#postfix_inet_protocols), where we suggest that you probably want IPV4.
13. We aren't using `systemd`, but we already have `supervisord` within the container managing services for you, Postfix is handled the same way, if the process is killed it'll be restarted, if the container is killed, it can likewise be restarted (via `restart: always` or similar policy in `docker-compose.yml`).

I'll stop there, perhaps you meant the entire article series with that reference?

I have not gone over the full series, some of it won't apply to this project obviously because of different choices. But I agree the structure can be helpful.

As illustrated above, we already handled the bulk of the Postfix setup as that article advised.

**Comment by VetheonGames** [comment]
Hey,

I do understand your expectation that I wouldn't actually read your response, but I promise I actually do. 

(Maybe not so much of the Tutorial stuff as I've already got my situation sorted, but I at least *look through* it)

And you are correct. I was more upset with the Docker than you or your guys' project and I came out the gate as a jerk and you guys really didn't deserve that. 

You guys are out here making mail servers more accessible to folks who haven't been a systems administrator for several years. I respect the *hell* out of that and you guys don't deserve some d**k-head being a jerk on the internet for no reason.

And I wasn't referring to the content directly in the tutorial series. But rather the structure and the flow of it. More the writing itself was what I referencing, not the content contained therein.

Also, I wouldn't suggest you make an LXC container, honestly. LXC is low-level containers and would probably be pretty insecure to run a mail server off of.

LXC provides a set of tools to manage your containers as well as templates to create a virtual environment of the most common Linux OS. Docker is an open-source containerization technology that focuses on running a single application in an isolated environment.

So as you can see, there *is* in fact several use cases here. However, if one knows *how* LXC works, one can make it function identically to Docker. The power with LXC is that it runs everything on a linux OS of your choice. So, for example, one could release a custom linux distro/config setup with LXC that automatically does the entire smtp/dovecot/etc installation and configuration for you.

But again, I wouldn't recommend doing that for mail servers. It would require *far* too much effort to do for free.

In terms of the docs, perhaps if I have some free time I can submit a PR or two to help update the docs to be more tutorial-y. I always like helping out open source developers. I can't do much to improve the image (other than maybe putting one of my image devs on it) but the docs I could certainly find some time to work on.

Again I have to say, I genuinely feel awful for the way I attacked you. I do sincerely apologize.

**Comment by polarathene** [comment]
> You guys are out here making mail servers more accessible to folks who haven't been a systems administrator for several years. I respect the _hell_ out of that and you guys don't deserve some d**k-head being a jerk on the internet for no reason.
> 
> Again I have to say, I genuinely feel awful for the way I attacked you. I do sincerely apologize.

Thanks :)

It's not an uncommon experience in the open-source world to see sometimes :sweat_smile:

I tend to ignore it, but I genuinely want this project to succeed and stay healthy. Good docs are important at acquiring a user base and retaining them, which can hopefully lead to more contributors. So I figured it was worth reaching out to know where exactly in the docs we could improve!

---

### LXC

> But again, I wouldn't recommend doing that for mail servers. It would require far too much effort to do for free.

Oh yeah, I have no intention of porting this project. It's quite large and I still don't have a handle over all of it just yet.

I think I'll give LXC a look for smaller personal projects in future though.

---

### Github Wiki vs `mkdocs-material` (our current docs solution)

> And I wasn't referring to the content directly in the tutorial series. But rather the structure and the flow of it. More the writing itself was what I referencing, not the content contained therein.

Yeah, I figured you hinting towards that, but I wasn't 100% sure.

> If you're wanting to improve the docs, first thing first, I would recommend you open a "wiki" section in the github repo.

I'm going to have to disagree with you there. We used to have that prior to early 2021, until [I merged the commit history of the wiki into our main repo](https://github.com/docker-mailserver/docker-mailserver/pull/1827) (_just to retain contributor attribution history as that matters to me_).

The separate wiki for this project wasn't working out. Yes it was easier for less experienced contributors to help with our docs, but they were an utter mess rotting with time, and no review process to maintain quality.

More importantly, the disconnect from actual features and project work in the main repo added friction for traceability (_which is quite helpful for me when trawling back in time with `git blame` to understand the history/decisions of parts of the project_).

We didn't just move the wiki markdown pages into the repo though. We added a doc generator (`mkdocs-material`) into the CI to give it a fresh slap of paint and better manage navigation. Links were corrected, pages re-organized and content revised. Revisions were mostly addressing outdated content, adopting new feature syntax (like admonitions, those little note/tip/warning boxes), fixing grammar/spelling... but no major overhaul over the content/structure. That process alone took a hefty chunk of time :grimacing: 

Here is some before/after:

![home_before](https://user-images.githubusercontent.com/5098581/172967170-0b590862-10fb-48ea-9a4d-3eff41094bfa.png)
![home_after](https://user-images.githubusercontent.com/5098581/172967191-59ad0a85-67d2-4de0-9ca3-32d3255c0aab.png)

![faq_before](https://user-images.githubusercontent.com/5098581/172967204-0226d171-a55d-4df9-af8c-fc70b8ca5166.png)
![faq_after](https://user-images.githubusercontent.com/5098581/172967220-ced2ebb3-0dcf-4265-a950-c1b3f78a014f.png)

![security-ports_before](https://user-images.githubusercontent.com/5098581/172967252-cdca98df-4b0b-4483-8b04-4088365d1f52.png)
![security-ports_after](https://user-images.githubusercontent.com/5098581/172967292-40efbeab-6ff5-49fc-879c-2b34bc846dad.png)

Now we have docs in the CI, previews are generated for PRs updating the docs and when merged, CI will deploy docs to the `gh-pages` branch.
- These docs are versioned, so it's possible to see the docs as they were for a past release which can be helpful, there's a little version drop-down at the top when viewing the docs (`edge` by default).
- All internal doc links should be raising a failure if a PR deletes doc content that was being linked from other doc pages (_useful when moving docs around or renaming pages as it catches some mistakes there which happened with the old wiki_).

There may have been other reasons/benefits, but the above is what comes to mind. If Github has improved their Wiki offering over the past year to address the issues we had, that's terrific. The current solution with `mkdocs-material` isn't exactly what I wanted, but was a fairly decent choice at the time and the other maintainers were more comfortable with it than alternatives I was also pitching.

---

### Contributions to docs

> It's much easier to format, plus it gives you additional pages and all kinds of more markdown/formatting options.

Anything specifically? I haven't kept up with the Wiki service, but I imagine we support most of what you're thinking about?

> In terms of the docs, perhaps if I have some free time I can submit a PR or two to help update the docs to be more tutorial-y. I always like helping out open source developers.

That'd be fantastic :+1: 

We are using [`mkdocs-material`](https://squidfunk.github.io/mkdocs-material/reference/) (syntax/feature reference page).

You can install that locally if you'd like to generate and preview docs. We have some light documentation regarding that [here](https://docker-mailserver.github.io/docker-mailserver/edge/contributing/documentation/). Our CI [makes a few minor changes](https://github.com/docker-mailserver/docker-mailserver/blob/e3cc627e18231b22cad715111b5fcf00703e08d7/.github/workflows/docs-production-deploy.yml#L45) that you don't have to worry about locally. It runs [this docker command](https://github.com/docker-mailserver/docker-mailserver/blob/e3cc627e18231b22cad715111b5fcf00703e08d7/.github/workflows/scripts/docs/build-docs.sh#L4-L13) to build docs, otherwise the command shown in our docs should provide a dev server IIRC that rebuilds docs automatically as you make changes.

Alternatively, you can skip all that and just work on the markdown docs without viewing the rendered version. Once you submit a PR we'll generate a temporary build with your changes and you can double check that it rendered correctly there if you like.

Our docs are in markdown format [here](https://github.com/docker-mailserver/docker-mailserver/tree/master/docs/content), with the extended syntax I linked for `mkdocs-material` when you see anything unfamiliar. And navigation hierarchy for pages is managed [here in `mkdocs.yml`](https://github.com/docker-mailserver/docker-mailserver/blob/e3cc627e18231b22cad715111b5fcf00703e08d7/docs/mkdocs.yml#L111). If you have any questions let me know :grinning: 

---

### Contributions beyond docs

> I can't do much to improve the image (other than maybe putting one of my image devs on it) but the docs I could certainly find some time to work on.

Contributing to the image is a bit more difficult but welcome.

Usually we get contributions from users trying to fix a problem they are facing or add a feature they want. That is welcome, just know that I am more of a strict reviewer than we've had in the past. If the contributor is not willing to put the time in to alleviate concerns I raise and address feedback for a quality contribution, they tend to abandon the PR.

If they're comfortable with bash, helping refactor our scripts to be easier to maintain is great. Plenty of that has been covered over the past couple years already, our tests need similar attention, but often that requires more domain knowledge not just in our project but mail-servers in general.

The `Dockerfile`, we could probably optimize into a multi-stage image. And our CI pipeline with Github Actions needs someone to look at, as there's a job that's building the images for multi-platform as a prerequisite to our `x86`-only test suite run, this is bad since there's 20 minutes roughly spent on building images that aren't under test, followed by the tests themselves taking 20-30 minutes. It'd be great to refactor our test suite to at least with the CI split out into parallel runs, the ROI isn't too high however as usually we don't have a high frequency of PRs or running of tests, just a nice to have for DX.

Tests themselves, it'd be nice to dockerize (_presently you need to clone the project and initialize the submodule `bats`, and deal with cleanup scripts for files written to the host_). [This was investigated](https://github.com/docker-mailserver/docker-mailserver/issues/2027), but is a bit nuanced due to our project being a docker container to test against.

LDAP support could do with love too, due to outstanding issues.

Any help with that stuff is appreciated, but we'd be just as happy with improving the docs :+1:

**Comment by VetheonGames** [comment]
> I'm going to have to disagree with you there. We used to have that prior to early 2021, until https://github.com/docker-mailserver/docker-mailserver/pull/1827 (just to retain contributor attribution history as that matters to me).

I actually was not aware that you were using another service for docs. I genuinely didn't even think about that. I just assumed you were using a custom webpage lol

I have a pretty talented image dev who might be interested in working on this, I'll pass along the repo information to them.

Otherwise, thanks for reaching out to me, and thank you for being kind, even though I was not.
